src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs

```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
+                        }
+
+                        var task = InvokeNextResultFilterAsync();
+                        if (task.Status != TaskStatus.RanToCompletion)
+                        {
+                            next = State.ResultSyncEnd;
+                            return task;
+                        }
+
+                        goto case State.ResultSyncEnd;
+                    }
+
+                case State.ResultSyncEnd:
+                    {
+                        Debug.Assert(_resultExecutedContext != null);
+
+                        var filter = (IResultFilter)state;
+                        var resultExecutedContext = _resultExecutedContext;
 
-                        item.Filter.OnResourceExecuted(_resourceExecutedContext);
+                        _diagnosticSource.BeforeOnResultExecuted(resultExecutedContext, filter);
 
-                        _diagnosticSource.AfterOnResourceExecuted(_resourceExecutedContext, item.Filter);
+                        filter.OnResultExecuted(resultExecutedContext);
+
+                        _diagnosticSource.AfterOnResultExecuted(resultExecutedContext, filter);
+
+                        goto case State.ResultEnd;
                     }
-                }
-                else
-                {
-                    // >> ExceptionFilters >> Model Binding >> ActionFilters >> Action
-                    await InvokeAllExceptionFiltersAsync();
 
-                    // If Exception Filters provide a result, it's a short-circuit due to an exception.
-                    // We don't execute Result Filters around the result.
-                    Debug.Assert(_exceptionContext != null);
-                    if (_exceptionContext.Result != null)
+                case State.ResultInside:
                     {
-                        // This means that exception filters returned a result to 'handle' an error.
-                        // We're not interested in seeing the exception details since it was handled.
-                        await InvokeResultAsync(_exceptionContext.Result);
+                        // The empty result is always flowed back as the 'executed' result
+                        var result = _resultExecutingContext?.Result ?? _result ?? new EmptyResult();
+                        _result = result;
 
-                        _resourceExecutedContext = new ResourceExecutedContext(_resourceExecutingContext, _filters)
+                        var task = InvokeResultAsync(_result);
+                        if (task.Status != TaskStatus.RanToCompletion)
                         {
-                            Result = _exceptionContext.Result,
-                        };
+                            next = State.ResultEnd;
+                            return task;
+                        }
+
+                        goto case State.ResultEnd;
                     }
-                    else if (_exceptionContext.Exception != null)
+
+                case State.ResultEnd:
```

Comment:
Please take a look at what's there and add specific comments. I've added comments in a few places already so if you think something specific is missing point it out.